### PR TITLE
Allow arrow function parameter parsing to bail out during speculation

### LINKED ISSUE #32914: Unexpected parse problems in valid JS code when arrow functions are used as values in object surrounded with parentheses
**TypeScript Version:**  3.7.0-dev.20190815 (there are same errors in playground even in 2.x)

**Search Terms:** Identifier expected , arrow function in object

**Code**

`test.js`
```js
const value = 1;
const a = {
    b: () => false
}

const test = () => ({
    // "Identifier expected." error on "!" and two "Duplicate identifier '(Missing)'." errors on space.
    prop: !value, // remove ! to see that errors will be gone
    run: () => { //replace arrow function with regular function to see that errors will be gone
        // comment next line or remove "()" to see that errors will be gone
        if(!a.b()) { return 'special'; }

        return 'default';
    }
});
```
There are no look-alike Unicode characters.
There are no problems if you switch order of `prop` and `run`.

Error is reproducible even when I write
```js
const test = ({
    prop: !value
    run: () => {
        if(!a.b()) { return 'special'; }

        return 'default';
    },
});
```
But works as expected if I remove parentheses around object or replace arrow function with regular one.

**Expected behavior:** No errors (code is valid and produces expected results in browsers)

**Actual behavior:** A lot of errors, most notably `Identifier expected.` on `!` character in line `prop: !value`

**UPD.** Object properties are actually parsed as arrow function parameters:
image

**Playground Link:** [Example](https://www.typescriptlang.org/play/index.html?noImplicitAny=false&target=6#code/MYewdgzgLgBAbgQwDYFcCmMC8MCMBuAKFElgSxgG8CYaYAjALhgAoBKLAPhgDNkI0CAX0JFw0GFDTQc5NpxZVaMAPTKYAIgCSAEzRgoAS24G0AJxhoAHgAc0wSdoB06i6dMhz4DQEIXCMNoSAO4gGgAiKNZIBsAIkjAGuvpGJuYA5MwAsgYQEAZgAOasac6u7qYQMF4Q1gjAaI7UtNbu1kzeiKhoADQqaqZoALYgcBjeEqH8GFAAFnFlHpVBBkhI9BgF4AJKpihgTHKYXBR9A1F1GAhuIEE8e-YGXsuzMAMFKEhXd2APXlCTaGmc1gZnKSxWazoGy2TSUNFUMFAg0GelgYCssGi6Kq5gGw1GGjYLn+MCmEmBCwqMGWq3WME26NhcKMzG8CEcdDY7BOAygKFMYBgaRqdgMyDSeBgggITJ2aD5AqFul4HygEqZgm6QlYhCAA) (I disabled `noImplicitAny` to remove noise)

**Related Issues:** Not found

### PR DESCRIPTION
**Pull request by weswigham**
Fixes #32914

### CODE REVIEW

**Comment by orta** [comment]
Impressive, almost every line of the change is now a conflict 👍🏻

File: src/compiler/parser.ts [hunk 1]
```diff
@@ -3974,7 +3999,7 @@ namespace ts {
             // a => (b => c)
             // And think that "(b =>" was actually a parenthesized arrow function with a missing
             // close paren.
-            if (!fillSignature(SyntaxKind.ColonToken, isAsync, node) && !allowAmbiguity) {
+            if (!fillSignature(SyntaxKind.ColonToken, isAsync, node, allowAmbiguity) && !allowAmbiguity) {
```
Comment (orta) [inline src/compiler/parser.ts:4002]: I'm looking into rebasing this PR, I got reasonably far, but all of the signature stuff has been completely removed from the codebase - so I'm going to wait on this till our pairing session

**Comment by jakebailey** [comment]
@weswigham Do you want to pick this PR up again, or just close it? I also tried to sit down to rebase/merge main back in, but wasn't very successful. Might just need a redo (or you to fix it if you know the intent behind some of the changes).

**Comment by jakebailey** [comment]
I'm going to close this in favor of #48493, if that's alright. This was a bit too hard to rebase on the branch so I did it by hand against `main`'s parser.

**Pull request closed without merging**